dotnet/src/InternalUtilities/connectors/AI/FunctionCallsProcessor.cs

```diff
@@ -0,0 +1,342 @@
+                    // further calls made as part of this function invocation. In particular, we must not use function calling settings naively here,
+                    // as the called function could in turn telling the model about itself as a possible candidate for invocation.
+                    context.Result = await function.InvokeAsync(kernel, invocationContext.Arguments, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }).ConfigureAwait(false);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception e)
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                this.AddFunctionCallResultToChatHistory(chatHistory, functionCall, result: null, errorMessage: $"Error: Exception while invoking function. {e.Message}");
+                continue;
+            }
+            finally
+            {
+                s_inflightAutoInvokes.Value--;
+            }
+
+            // Apply any changes from the auto function invocation filters context to final result.
+            functionResult = invocationContext.Result;
+
+            object functionResultValue = functionResult.GetValue<object>() ?? string.Empty;
+
+            var result = ProcessFunctionResult(functionResultValue, resultSerializerOptions);
+
+            this.AddFunctionCallResultToChatHistory(chatHistory, functionCall, result);
+
+            // If filter requested termination, return last chat history message.
+            if (invocationContext.Terminate)
+            {
+                if (this._logger.IsEnabled(LogLevel.Debug))
+                {
+                    this._logger.LogDebug("Filter requested termination of automatic function invocation.");
+                }
+
+                return chatHistory.Last();
+            }
+        }
+
+        return null;
+    }
+
+    /// <summary>
+    /// Adds the function call result or error message to the chat history.
+    /// </summary>
+    /// <param name="chatHistory">The chat history to add the function call result to.</param>
+    /// <param name="functionCall">The function call.</param>
+    /// <param name="result">The function result to add to the chat history.</param>
+    /// <param name="errorMessage">The error message to add to the chat history.</param>
+    private void AddFunctionCallResultToChatHistory(ChatHistory chatHistory, FunctionCallContent functionCall, string? result, string? errorMessage = null)
+    {
+        // Log any error
+        if (errorMessage is not null && this._logger.IsEnabled(LogLevel.Debug))
+        {
+            Debug.Assert(result is null);
+            this._logger.LogDebug("Failed to handle function request ({Id}). {Error}", functionCall.Id, errorMessage);
+        }
+
+        result ??= errorMessage ?? string.Empty;
```

Comment:
The `errorMessage` will go back to the LLM so we need to make sure it contain sufficient information for it to respond appropriately.

Reply:
Do we have examples of inappropriate behavior that we can analyze to understand how to address it?

Reply:
Captured in this issue: https://github.com/microsoft/semantic-kernel/issues/8490, to be addressed later.